Objects/setobject.c

```diff
@@ -2773,17 +2773,23 @@ PySet_Discard(PyObject *set, PyObject *key)
 int
 PySet_Add(PyObject *anyset, PyObject *key)
 {
-    if (!PySet_Check(anyset) &&
-        (!PyFrozenSet_Check(anyset) || !_PyObject_IsUniquelyReferenced(anyset))) {
-        PyErr_BadInternalCall();
-        return -1;
+    if (_PyObject_IsUniquelyReferenced(anyset) && PyAnySet_Check(anyset)) {
+        // In free-threading, if the set or frozenset is uniquely referenced,
+        // no critical section is needed since only the owner thread is
+        // populating it.
```

Comment:
Let's make it obvious that this covers frozensets (the difference between PySet_Check and PyAnySet_Check is too easy to miss otherwise.)

```suggestion
        // We can only change frozensets if they are uniquely referenced,
        // and we can avoid locking sets even in free-threaded build if they
        // are uniquely referenced.
```

Reply:
I was not aware that the public `PySet_Add` can modify frozensets. This interferes with https://github.com/python/cpython/pull/140234, where we skip tracking of `frozenset` objects containing immutables by the GC. If this PR is merged earlier, I will adapt the other PR.

Reply:
`PySet_Add` was able to modify frozensets before this PR. That hasn't changed (see https://github.com/python/cpython/commit/cab3d98ca19659698a29108aabb040f2741d1a2a)